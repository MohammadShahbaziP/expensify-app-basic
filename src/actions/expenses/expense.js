
import database from '../../firebase/firebase'

const removeExpense = ({id} = {})=>({
    type:'REMOVE_EXPENSE',
    id
})

const startRemoveExpense = (id)=>{
    console.log(id)
    return(dispatch,getState)=>{
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}))
        })
    }
}

const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})


export const startAddExpense = (expenses ={})=>{
    return (dispatch, getState)=>{
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenses

        const expense = {description , note, amount ,createdAt}
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

const editExpense= (id, update)=>({
    type:'EDIT_EXPENSE',
    id,
    update
})

const startEditExpense = (id , update)=>{
    return(dispatch,getState)=>{
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`).update({
            ...update
        }).then(()=>{
            dispatch(editExpense(id,update))
        })
    }
}

const setExpense =(expenses)=>({
    type:'SET_EXPENSES',
    expenses
})

const startSetExpense=()=>{
    return (dispatch,getState)=>{
        const uid = getState().auth.uid
        
        return database.ref(`users/${uid}/expenses`).once('value').then((snapsht)=>{
            const expenses=[]
            snapsht.forEach((childSnapshot)=>{
                expenses.push({
                    id:childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpense(expenses))
        })        
    }
}
export {addExpense, startEditExpense, removeExpense, editExpense, setExpense ,startSetExpense, startRemoveExpense}