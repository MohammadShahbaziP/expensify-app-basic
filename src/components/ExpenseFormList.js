import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)
        this.state={
            description:props.expense? props.expense.description :'',
            note:props.expense? props.expense.note :'',
            amount:props.expense? (props.expense.amount / 100).toString() :'',
            createdAt:props.expense? moment(this.props.amount) :moment(),
            calendareFocused:false,
            error:''
        }
    }

    handelChangeDescription = (e)=>{
        const description = e.target.value
        this.setState(()=>({ description }))

    }
    handelChangeNote = (e)=>{
        const note = e.target.value
        this.setState(()=>({ note }))

    }
    onAmountChange= (e)=>{
        const amount = e.target.value
        if(amount.match(/^\d*(\.\d{0,2})?$/)){
            this.setState(()=>({ amount }))
        }
    }
    onDateChange= (createdAt)=>{
        if (createdAt){
            this.setState(()=>({ createdAt }))            
        }

    }
    onFocusChange= ({ focused })=>{
        this.setState(()=>({ calendareFocused:focused}))
    }
    onSubmit= (e)=>{
        e.preventDefault()

        if(!this.state.amount && !this.state.description){
            this.setState({error: 'please fill the amount and description' })
        }else{
            this.setState({error: '' })
            this.props.onSubmit({
                description:this.state.description,
                note:this.state.note,
                amount:parseFloat(this.state.amount) *100,
                createdAt:this.state.createdAt.valueOf()
            })
        }
    }
    render(){
        return(
        <div>
        {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit} >
                <input
                    type='text'
                    autoFocus
                    placeholder='Description'
                    value={this.state.description}
                    onChange={this.handelChangeDescription}
                />
                <input 
                    type='text'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    placeholder='Amount'
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendareFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                />
                <textarea 
                    placeholder='Note'
                    value={this.state.value}
                    onChange={this.handelChangeNote}
                />
                <button>Add expense</button>
            </form>
        </div>
        )
    }

}