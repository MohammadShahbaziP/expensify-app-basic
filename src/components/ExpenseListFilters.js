import React from 'react'
import { connect } from 'react-redux'
import {setFilterText , sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters/filter'
import { DateRangePicker} from 'react-dates'
export class ExpenseListFilters extends React.Component {
    state = {
        focusedCallender:null
    }
    onDatesChange=({startDate , endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange=(focusedInput)=>{
        this.setState(()=>({focusedCallender:focusedInput}))

    }
    onTextChange= (e)=>{
        this.props.setFilterText(e.target.value)
    }

    onSortChange = (e)=>{
        e.target.value === 'date' ? this.props.sortByDate():this.props.sortByAmount()
    }
    render() {
        return(
            <div>
            <input type='text' value={this.props.filters.text} onChange={this.onTextChange}/>
            <select value={this.props.filters.sortBy} onChange={this.onSortChange}
            >
            <option value='date'>Date</option>
            <option value='amount'>Amount</option>
            
            </select>
            <DateRangePicker 
                startDate={this.props.filters.startAt}
                endDate={this.props.filters.endAt}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.focusedCallender}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                showClearDates={true}
            />

        </div>
        )
    }
}

const mapDispatchToProps = (dispatch) =>({
    setFilterText : (text)=>dispatch(setFilterText(text)),
    setStartDate : (startDate)=>dispatch(setStartDate(startDate)),
    setEndDate : (endDate)=>dispatch(setEndDate(endDate)),
    sortByDate : (date)=> dispatch(sortByDate(date)),
    sortByAmount : (amount)=>dispatch(sortByAmount(amount))
})

const mapStateToProps = (state) =>({
    filters:state.filters
})
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters) 