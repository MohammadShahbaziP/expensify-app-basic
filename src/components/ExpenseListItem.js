import React from 'react'
import {connect} from 'react-redux'
import { Link } from 'react-router-dom'

 const ExpenseListItem = (props)=>(
    <div>
        <Link  to={'/edit/'+props.expense.id} >{props.expense.description}</Link>
        <p>{props.expense.amount}</p>
        
        

    </div>
)
export default ExpenseListItem