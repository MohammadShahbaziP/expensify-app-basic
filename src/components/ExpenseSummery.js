import React from 'react'
import { connect } from 'react-redux'
import getTotalAmount from '../selectors/getTotalAmount'
import getVisibleState from '../selectors/getVisiblestate'
import numeral from 'numeral'
export const ExpenseSummery = (props) =>{
    const FilteredExpenses = getVisibleState(props.expenses, props.filters)
    const total = getTotalAmount(FilteredExpenses)
    const expensesWord= FilteredExpenses.length === 1 ? 'expense' : 'expenses'
    return(
        <div>
            <p>Viewing {FilteredExpenses.length} {expensesWord} totaling {numeral(total / 100).format('$0,0.00')}</p>
        </div>
    )
}

const mapStatetoProps= (state)=>({
    expenses : state.expenses,
    filters: state.filters
})
export default connect(mapStatetoProps)(ExpenseSummery )