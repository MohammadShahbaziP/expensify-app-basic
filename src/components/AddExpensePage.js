import React from 'react'
import { connect } from 'react-redux'
import { startAddExpense } from '../actions/expenses/expense'
import ExpenseForm from './ExpenseFormList'

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <div className="summery">
                    <div className="container">
                        <h1 className="summery__title">Add expense </h1>
                    </div>


                </div>
                <div className="container">
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>

            </div>
        )

    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(startAddExpense(expense))
})
export default connect(undefined, mapDispatchToProps)(AddExpensePage)