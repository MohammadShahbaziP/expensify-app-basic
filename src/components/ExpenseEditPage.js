import React from "react"
import { connect } from 'react-redux'
import { editExpense } from '../actions/expenses/expense'
import ExpenseForm from './ExpenseFormList'
import {startRemoveExpense} from '../actions/expenses/expense'

export class ExpenseEditPage extends React.Component{
    onSubmit=(expense)=>{
        this.props.editExpense(this.props.expense.id,expense)
        this.props.history.push('/')
    }
    onClick=()=>{
        this.props.removeExpense({id:this.props.expense.id})
        this.props.history.push('/')
    }
    render(){
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>remove</button>
            
            </div>
            )
    }
}
 

const mapDispatchToProps= (dispatch)=>({
    removeExpense: ({id})=>dispatch(startRemoveExpense(id)),
    editExpense: (id,expense)=>dispatch(editExpense(id,expense))
})
const mapStateToProps = (state, props)=>{
    return {
        expense: state.expenses.find((expense)=>{
            return expense.id===props.match.params.id
        })
    }
}
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseEditPage)