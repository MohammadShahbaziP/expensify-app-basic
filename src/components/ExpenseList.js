import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpense from '../selectors/getVisiblestate'

export const ExpenseList = ({expenses})=>(
    <div>

        {expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
           
             expenses.map((expense)=>(
                <ExpenseListItem key={expense.id} expense={expense}  />
            ))
            
        )}

    
    </div>
) 
const mapStatetoProps = (state)=>{
    return {
        expenses:getVisibleExpense(state.expenses, state.filters)
    }
}
export default connect(mapStatetoProps)(ExpenseList)