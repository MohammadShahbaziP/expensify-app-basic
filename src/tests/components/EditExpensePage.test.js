import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseEditPage } from '../../components/ExpenseEditPage'
import expenses from '../fixturs/expenseList'
let editExpense, removeExpense,history, wrapper,expense

beforeEach(()=>{
    editExpense=jest.fn()
    removeExpense=jest.fn()
    history={push:jest.fn()}
    expense=expenses[0]
    wrapper=shallow(<ExpenseEditPage expense={expense} editExpense={editExpense} removeExpense={removeExpense} history={history} />)

})

test('should render expense edit page correctly',()=>{
    expect(wrapper).toMatchSnapshot()
})
test('should on submit expense form',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,expenses[0])
})

test('should on click remove button',()=>{
    const id = expenses[0].id
    wrapper.find('button').simulate('Click')
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(removeExpense).toHaveBeenLastCalledWith({id})
})