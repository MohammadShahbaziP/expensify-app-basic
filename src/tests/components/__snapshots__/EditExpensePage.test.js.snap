// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense edit page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 200,
        "createdAt": 0,
        "description": "gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
